gprt/gprt.cpp

```diff
@@ -1478,40 +1495,65 @@ struct AABBAccel : public Accel {
 };
 
 struct InstanceAccel : public Accel {
+  size_t numInstances;
   std::vector<Accel*> instances; 
 
+  // the total number of geometries referenced by this instance accel's BLASes
+  size_t numGeometries = -1;
+
   size_t instanceOffset = -1;
 
   Buffer *instancesBuffer = nullptr;
   Buffer *accelAddressesBuffer = nullptr;
+  Buffer *instanceOffsetsBuffer = nullptr;
+
+  struct {
+    Buffer* buffer = nullptr;
+    size_t stride = 0;
+    size_t offset = 0;
+  } transforms;
 
   struct {
     Buffer* buffer = nullptr;
     // size_t stride = 0;
     // size_t offset = 0;
-  } transforms;
-  
+  } references;
+
+  struct {
+    Buffer* buffer = nullptr;
+    // size_t stride = 0;
+    // size_t offset = 0;
+  } offsets;
+
   // todo, accept this in constructor
   VkBuildAccelerationStructureFlagsKHR flags = VK_BUILD_ACCELERATION_STRUCTURE_PREFER_FAST_TRACE_BIT_KHR;
 
   InstanceAccel(VkPhysicalDevice physicalDevice, VkDevice logicalDevice, VkCommandBuffer commandBuffer, VkQueue queue,
     size_t numInstances, GPRTAccel* instances) : Accel(physicalDevice, logicalDevice, commandBuffer, queue) 
   {
-    this->instances.resize(numInstances);
-    memcpy(this->instances.data(), instances, sizeof(GPRTAccel*) * numInstances);
-
-    accelAddressesBuffer = new Buffer(
-      physicalDevice, logicalDevice, commandBuffer, queue,
-      // I guess I need this to use these buffers as input to tree builds?
-      VK_BUFFER_USAGE_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT_KHR | 
-      // means we can get this buffer's address with vkGetBufferDeviceAddress
-      VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT, 
-      // means that this memory is stored directly on the device 
-      //  (rather than the host, or in a special host/device section)
-      // VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT | // temporary (doesn't work on AMD)
-      VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT, // temporary
-      sizeof(uint64_t) * numInstances
-    );
+    this->numInstances = numInstances;
+
+    if (instances) {
+      this->instances.resize(numInstances);
+      memcpy(this->instances.data(), instances, sizeof(GPRTAccel*) * numInstances);
```

Comment:
I'm a little concerned about memory leaks here. Will the `Accel` objects clean themselves up correctly if they're replaced in the vector?